shared-bindings/ulab/__init__.c

```diff
@@ -0,0 +1,637 @@
+/*
+ * This file is part of the micropython-ulab project,
+ *
+ * https://github.com/v923z/micropython-ulab
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019-2020 Zoltán Vörös
+*/
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "py/runtime.h"
+#include "py/binary.h"
+#include "py/obj.h"
+#include "py/objarray.h"
+
+#include "compat.h"
+#include "ndarray.h"
+#include "linalg.h"
+#include "vectorise.h"
+#include "poly.h"
+#include "fft.h"
+#include "numerical.h"
+
+//|
+//| :mod:`ulab` --- Manipulate numeric data similar to numpy
+//| ========================================================
+//|
+//| .. module:: ulab
+//|    :synopsis: Manipulat numeric data similar to numpy
+//|
+//| `ulab` is a numpy-like module for micropython, meant to simplify and
+//| speed up common mathematical operations on arrays. The primary goal was to
+//| implement a small subset of numpy that might be useful in the context of a
+//| microcontroller. This means low-level data processing of linear (array) and
+//| two-dimensional (matrix) data.
+//|
+//| `ulab` is adapted from micropython-ulab, and the original project's
+//| documentation can be found at
+//| https://micropython-ulab.readthedocs.io/en/latest/
+//|
+//| `ulab` is modeled after numpy, and aims to be a compatible subset where
+//| possible.  Numpy's documentation can be found at
+//| https://docs.scipy.org/doc/numpy/index.html
+//|
+//| .. contents::
+//|
+
+//| .. attribute:: __version__
+//|
+//| The closest corresponding version of micropython-ulab
+//|
+STATIC MP_DEFINE_STR_OBJ(ulab_version_obj, "0.26.7");
+
+//| ulab.array -- 1- and 2- dimensional array
+//| ==================================================
+//| .. class:: ulab.array(values, *, dtype=float)
+//|
+//|   :param sequence values: Sequence giving the initial content of the array.
+//|   :param dtype: The type of array values, ``int8``, ``uint8``, ``int16``, ``uint16``, or ``float``
+//|
+//|   The `values` sequence can either be a sequence of numbers (in which case a
+//|   1-dimensional array is created), or a sequence where each subsequence has
+//|   the same length (in which case a 2-dimensional array is created).
+//|
+//|   In many cases, it is more convenient to create an array from a function
+//|   like `zeros` or `linspace`.
+//|
+//|    .. method:: asbytearray()
+//|
+//|       Returns an `array.array` object that refers to the same memory.
+//|       This can be used with functions and methods that require "buffers",
+//|       such as `audiocore.RawSample`.
+//|
+//|       Despite the name, the byte array preserves the type information of
+//|       the array.
+//|
+//|       **WARNING:** `asbytearray` is a `ulab`-only method; it has no
+//|       equivalent in numpy.
```

Comment:
I'd be tempted to remove this in favor of optimizing `bytes(a)` where a is a ulab.array instead. It looks like this and rawsize are the only two extra things preventing array from being a strict subset of numpy's array.

Reply:
No, it's not really bytes(a) because it acts like an array of the same data type and it refers to the same storage.  However, maybe we could make memoryview() work on them which would mean we could remove asbytearray.  For some reason, memoryview doesn't take ulab arrays.

There are other differences that make this not a particularly strict subset of numpy.  
 * fft, ifft are in a different location
 * spectrum isn't in numpy
 * fft, ifft deal with complex values differently (numpy has arrays with complex elements, ulab does not)
 * flatten is marked as not conforming but I don't know why

Reply:
The existing name is confusing if it doesn't return a `bytearray`. I suggested bytes because it is immutable and we could refer to the same memory. That could be weird though if it is changed through the original array. memoryview is a better idea and I bet it works with np.array and cpython's memoryview.

I leave it up to you whether to move things around so there is a module that is a strict subset of numpy. (meaning ulab code can be run with numpy ok, not necessarily the reverse)

Reply:
But has `ulab` got to be a strict subset of `numpy`? I see a couple of reasons for a slight departure from `numpy`, and this is why I didn't call the library `unumpy` or something like that. 

1. `numpy` has at least 11 data types, perhaps more. It is extremely expensive to implement all those for binary operations, because the costs scale in a quadratic fashion with the number of types. It seems to me that the library is already too big in certain cases: https://github.com/v923z/micropython-ulab/issues/21#issuecomment-582648097. 
On the other hand, once you have to stipulate your own upcasting rules, you are not going to get the same results in `ulab` as in `numpy`, even if the code runs without any errors in both environments.

2. If you do any data processing, you might also need a couple of things from `scipy`. I wanted to have a single library for everything, so that setting up your compilation environment is not so involving.

Having said this, I am not against some sort of re-factoring, if you have convincing arguments, though, we should not forget that that will incur an increase in the firmware size. E.g., at the moment, `LinalgError` is not defined in `ulab`. In `numpy`, `linalg` is a sub-package, as is `fft`, as Jeff has pointed out. 

We should also keep in mind that `micropython` itself is not a strict subset of CPython, either. E.g., you won't find `const`s in CPython, yet, no one has ever complained  about it in the `micropython` community.

Reply:
@v923z I noticed that `.shape` is a property in numpy and a method in ulab.  Would you be receptive to changing this, or should we expect to carry such a change locally if we want it?

Reply:
> @v923z I noticed that `.shape` is a property in numpy and a method in ulab. Would you be receptive to changing this, 

Absolutely. That was the leftover of the very first version, and I just haven't got the time to fix it. I have the implementation, I will push the changes soon.

> or should we expect to carry such a change locally if we want it?

No, we shouldn't do that.